Source/DafnyLanguageServer/Handlers/DafnyHoverHandler.cs

```diff
@@ -215,15 +215,30 @@ class AssertionBatchIndexComparer : IComparer<AssertionBatchIndex> {
 
       string GetDescription(Boogie.ProofObligationDescription? description) {
         switch (assertionNode?.StatusVerification) {
-          case GutterVerificationStatus.Verified:
-            return $"{obsolescence}<span style='color:green'>**Success:**</span> " +
-                   (description?.SuccessDescription ?? "_no message_");
+          case GutterVerificationStatus.Verified: {
+              var successDescription = description?.SuccessDescription ?? "_no message_";
+              if (successDescription == "error is impossible: This is the precondition that might not hold.") {
+                successDescription = "the precondition always holds";
+              }
+              return $"{obsolescence}<span style='color:green'>**Success:**</span> " +
+                     successDescription;
+            }
           case GutterVerificationStatus.Error:
             var failureDescription = description?.FailureDescription ?? "_no message_";
             if (currentlyHoveringPostcondition &&
                   (failureDescription == new PostconditionDescription().FailureDescription ||
                    failureDescription == new EnsuresDescription().FailureDescription)) {
-              failureDescription = "This postcondition might not hold on a return path.";
+              failureDescription = "this postcondition might not hold on a return path";
+            }
+
+            if (!currentlyHoveringPostcondition &&
+                (failureDescription == new EnsuresDescription().FailureDescription)) {
+              failureDescription = "a postcondition could not be proven on this return path";
```

Comment:
Why does this message need to be changed? Why can't the back-end emit the right message to begin with? I think it would be great if DafnyHoverHandler can be completely Dafny agnostic.

Reply:
>Why does this message need to be changed? Why can't the back-end emit the right message to begin with? I think it would be great if DafnyHoverHandler can be completely Dafny agnostic.

Agree.
I created a new `EnsuresDescription` in Dafny so that we don't need this work-around Boogie's `EnsuresDescription`